### ISSUE TITLE
StdioOnnaStick ignores sys.stdout.encoding, won't print unicode properly

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1990]
|<img alt="termie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| termie reported|
|-|-|
|Trac ID|trac#1990|
|Type|defect|
|Created|2006-08-07 17:37:43Z|
|Branch|https://github.com/twisted/twisted/tree/stdio-unicode-1990|

Twisted overloads stdout to go to its log, but the log ignores encoding and tends to choke on unicode even though my terminal can support it

Attachments:

* [log.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb3/bb3acf149db4936fbaca693a61d56be89205d997/d0c2af7a4d0a496c6d546fa38d5b84158b4d865f.patch) (2246 bytes) - added by termie on 2006-08-07 17:38:39Z - A patch to have StdioOnnaStick default to the sys.stdout.encoding and push that through to msg()
* [print.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb3/bb3acf149db4936fbaca693a61d56be89205d997/ac3c0ea885293cad5410b05071ab3d75be3aac84.patch) (795 bytes) - added by termie on 2006-08-08 17:11:02Z - A patch fixing only the StdioOnnaStick side of things

<details><summary>Searchable metadata</summary>

```
trac-id__1990 1990
type__defect defect
reporter__termie termie
priority__normal normal
milestone__ 
branch__branches_stdio_unicode_1990 branches/stdio-unicode-1990
branch_author__itamarst__termie itamarst, termie
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1154972263000000 1154972263000000
changetime__1307103299000000 1307103299000000
version__None None
owner__itamar itamar
cc__spiv cc__exarkun cc__miracle2k cc__davidsarah
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to termie|
|-|-|

Needs tests.  I also don't think this is the right way to go about this.  StdioOnnaStick can encode any unicode it sees: the rest of the logging system doesn't need to know anything about this.

**Comment by twisted-trac** [comment]
|<img alt="termie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|termie commented|
|-|-|

glyph, safe_str is at fault here, too, as it by design only wants to handle ascii. So either safe_str should handle everything as unicode (which would be unsafe to most terminals and choke japanese), it should not be used, or it needs to have its encoding be 

As StdioOnnaStick has a known encoding, sys.stdout.encoding, it is safe for things going through there to use that encoding, however other logging applications do not have the benefit and as long as msg() uses safe_str, which fails on different encodings, and nothing accepts an encoding argument, nobody gets a chance to output proper data.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

If the purpose here is to fix `print', then my above comments apply.

If the purpose is to add unicode support to twisted.python.log, then there's a bit more work to do.  StdioOnnaStick isn't the right place to determine the encoding, since the log may not be going to stdout.  FileLogObserver does the actual write, so it should do the actual encoding.  safe_str shouldn't know about encodings, but maybe it should return a unicode string or there should be a safe_unicode.  The caller can worry about doing the encoding.

This all has the benefit of avoiding mixing byte strings with different encodings and passing an explicit encoding name through the logging system.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

The patch should include unit tests.  Thanks.

**Comment by twisted-trac** [comment]
|<img alt="davidsarah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|davidsarah commented|
|-|-|

The absence of an ```encoding``` attribute on ```StdioOnnaStick``` instances caused [http://tahoe-lafs.org/trac/tahoe-lafs/ticket/1099 Tahoe-LAFS bug [#10846](https://github.com/twisted/twisted/issues/10846)].

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

(In [32003]) Branching to 'stdio-unicode-1990'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst removed owner|
|-|-|

I wrote a bunch of tests for StdioOnnaStick, and then added the encoding attribute as well as code to encode unicode. The limited goal is to make print continue to work in the face of Twisted logging; if printing unicode used to work, it should continue to work.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @itamarst|
|-|-|

1. You should use addCleanup in the tests, instead of try/finally.

2. It would be nice to use better variable names than f and s.

3. You could reduce the number of calls to sys.getdefaultencoding by making the default encoding argument to `StdioOnnaStick.__init__` None, and use getdefaultencoding in that case.

Build results here: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/stdio-unicode-1990

Thanks, please merge!

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_12"></a>|@itamarst set status to `closed`|
|-|-|

(In [32022]) Merge stdio-unicode-1990

Author: termie, itamarst
Reviewer: therve
Fixes: [#1990](https://github.com/twisted/twisted/issues/1990)

When the log system overrides sys.stdout and sys.stderr, continue to support standard Python behavior of encoding Unicode in print.
